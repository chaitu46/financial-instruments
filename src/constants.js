export const COLORS = {
  blue: "#1d4289",
  red: "#d3273e",
  green: "#007a78",
  white: "white",
  black: "black",
  darkGrey: "#303030",
};

export const ASSERT_ID = "assertClass";
export const TICKER_ID = "ticker";
export const PRICE_ID = "price";

export const COLUMN_NAMES = ["Asset Class", "Ticker", "Price"];
export const COLUMN_IDS = [ASSERT_ID, TICKER_ID, PRICE_ID];
export const ASSERT_CUSTOM_ORDER = ["Equities", "Macro", "Credit"];

export const ROW_THEME = {
  Macro: {
    bg: COLORS.white,
  },
  Equities: {
    bg: COLORS.blue,
    color: COLORS.white,
  },
  Credit: {
    bg: COLORS.green,
    color: COLORS.white,
  },
};
