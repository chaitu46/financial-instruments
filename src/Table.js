import React, { useMemo, useState } from "react";
import {
  StyledHeaderButton,
  StyledHeadRow,
  StyledPriceTd,
  StyledTable,
  StyledTr,
} from "./styledComponents";
import { COLORS, COLUMN_IDS, COLUMN_NAMES, ROW_THEME } from "./constants";
import getSortData from "./calculations/getSortData";

export const Table = ({ initialData }) => {
  const [sortByColumn, setSortByColumn] = useState();
  const tableData = useMemo(() => getSortData(sortByColumn, initialData), [
    sortByColumn,
    initialData,
  ]);
  const handleSort = ({ currentTarget: { id } }) => {
    if (id !== sortByColumn) {
      setSortByColumn(id);
    }
  };

  return (
    <StyledTable>
      <caption>Financial details</caption>
      <thead>
        <StyledHeadRow>
          {COLUMN_NAMES.map((name, index) => (
            <th key={name}>
              <StyledHeaderButton
                onClick={handleSort}
                data-active={COLUMN_IDS[index] === sortByColumn}
                id={COLUMN_IDS[index]}
              >
                {name}
              </StyledHeaderButton>
            </th>
          ))}
        </StyledHeadRow>
      </thead>
      <tbody>
        {tableData.map((instrument, index) => (
          <StyledTr key={index} data-theme={ROW_THEME[instrument.assetClass]}>
            <td>{instrument.assetClass}</td>
            <td>{instrument.ticker}</td>
            <StyledPriceTd
              data-bg={instrument.price >= 0 ? COLORS.blue : COLORS.red}
            >
              {instrument.price}
            </StyledPriceTd>
          </StyledTr>
        ))}
      </tbody>
    </StyledTable>
  );
};
