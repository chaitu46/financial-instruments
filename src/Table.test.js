import { fireEvent, render, screen } from "@testing-library/react";
import { Table } from "./Table";
import { ASSERT_ID, COLUMN_NAMES, PRICE_ID, TICKER_ID } from "./constants";

const testData = [
  {
    ticker: "BETA",
    price: 3791.37,
    assetClass: "Equities",
  },
  {
    ticker: "EPSILON",
    price: 1168.46,
    assetClass: "Credit",
  },
  {
    ticker: "EPSILON",
    price: -1096.64,
    assetClass: "Macro",
  },
  {
    ticker: "GAMMA",
    price: 2299.1,
    assetClass: "Equities",
  },
];

describe("Table", () => {
  test("renders default Table", () => {
    render(<Table initialData={testData} />);
    const table = screen.getByRole("table");
    const headerButtons = screen.getAllByRole("button");

    expect(table).toBeVisible();
    expect(table).toHaveAttribute("aria-label", "Financial Details");
    expect(headerButtons).toHaveLength(3);
    headerButtons.map((button, index) =>
      expect(button).toHaveTextContent(COLUMN_NAMES[index])
    );
    expect(screen.getByTestId(ASSERT_ID)).toHaveAttribute(
      "data-active",
      "false"
    );
    expect(screen.getByTestId(PRICE_ID)).toHaveAttribute(
      "data-active",
      "false"
    );
    expect(screen.getByTestId(TICKER_ID)).toHaveAttribute(
      "data-active",
      "false"
    );
    expect(table).toMatchSnapshot();
  });

  test("simulate asset class column sort click", () => {
    render(<Table initialData={testData} />);
    const table = screen.getAllByRole("table");
    expect(table).toMatchSnapshot();
    fireEvent.click(screen.getByTestId(ASSERT_ID));
    expect(table).toMatchSnapshot();
    expect(screen.getByTestId(ASSERT_ID)).toHaveAttribute(
      "data-active",
      "true"
    );
    expect(screen.getByTestId(PRICE_ID)).toHaveAttribute(
      "data-active",
      "false"
    );
    expect(screen.getByTestId(TICKER_ID)).toHaveAttribute(
      "data-active",
      "false"
    );
  });
  test("simulate ticker column sort click", () => {
    render(<Table initialData={testData} />);
    const table = screen.getAllByRole("table");

    expect(table).toMatchSnapshot();
    fireEvent.click(screen.getByTestId(TICKER_ID));
    expect(table).toMatchSnapshot();
    expect(screen.getByTestId(ASSERT_ID)).toHaveAttribute(
      "data-active",
      "false"
    );
    expect(screen.getByTestId(PRICE_ID)).toHaveAttribute(
      "data-active",
      "false"
    );
    expect(screen.getByTestId(TICKER_ID)).toHaveAttribute(
      "data-active",
      "true"
    );
  });
  test("simulate price column sort click", () => {
    render(<Table initialData={testData} />);
    const table = screen.getAllByRole("table");

    expect(table).toMatchSnapshot();
    fireEvent.click(screen.getByTestId(PRICE_ID));
    expect(table).toMatchSnapshot();
    expect(screen.getByTestId(ASSERT_ID)).toHaveAttribute(
      "data-active",
      "false"
    );
    expect(screen.getByTestId(PRICE_ID)).toHaveAttribute("data-active", "true");
    expect(screen.getByTestId(TICKER_ID)).toHaveAttribute(
      "data-active",
      "false"
    );
  });

  test("simulate price column sort twice and ensure nothing changed", () => {
    render(<Table initialData={testData} />);
    fireEvent.click(screen.getByTestId(PRICE_ID));
    expect(screen.getByTestId(ASSERT_ID)).toHaveAttribute(
      "data-active",
      "false"
    );
    expect(screen.getByTestId(PRICE_ID)).toHaveAttribute("data-active", "true");
    expect(screen.getByTestId(TICKER_ID)).toHaveAttribute(
      "data-active",
      "false"
    );
    const mockGetSortByColumnName = jest.fn();
    jest.mock("./utils", () => ({
      getSortDataByColumnName: () => mockGetSortByColumnName,
    }));
    fireEvent.click(screen.getByTestId(PRICE_ID));
    expect(mockGetSortByColumnName).not.toBeCalled();
  });
});
