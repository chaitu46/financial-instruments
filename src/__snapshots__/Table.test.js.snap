// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table renders default Table 1`] = `
<table
  aria-label="Financial Details"
  class="sc-bdfBwQ keewPp"
>
  <thead>
    <tr
      class="sc-hKgILt eBPHrI"
    >
      <th>
        <button
          class="sc-eCssSg gYaMmc"
          data-active="false"
          data-testid="assertClass"
          id="assertClass"
        >
          Asset Class
        </button>
      </th>
      <th>
        <button
          class="sc-eCssSg gYaMmc"
          data-active="false"
          data-testid="ticker"
          id="ticker"
        >
          Ticker
        </button>
      </th>
      <th>
        <button
          class="sc-eCssSg gYaMmc"
          data-active="false"
          data-testid="price"
          id="price"
        >
          Price
        </button>
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      class="sc-gsTCUz kHDUxD"
      data-theme="[object Object]"
    >
      <td>
        Equities
      </td>
      <td>
        BETA
      </td>
      <td
        class="sc-dlfnbm eWIIxE"
        data-bg="#1d4289"
      >
        3791.37
      </td>
    </tr>
    <tr
      class="sc-gsTCUz ibRWgQ"
      data-theme="[object Object]"
    >
      <td>
        Credit
      </td>
      <td>
        EPSILON
      </td>
      <td
        class="sc-dlfnbm eWIIxE"
        data-bg="#1d4289"
      >
        1168.46
      </td>
    </tr>
    <tr
      class="sc-gsTCUz jmeBcN"
      data-theme="[object Object]"
    >
      <td>
        Macro
      </td>
      <td>
        EPSILON
      </td>
      <td
        class="sc-dlfnbm kxGDBW"
        data-bg="#d3273e"
      >
        -1096.64
      </td>
    </tr>
    <tr
      class="sc-gsTCUz kHDUxD"
      data-theme="[object Object]"
    >
      <td>
        Equities
      </td>
      <td>
        GAMMA
      </td>
      <td
        class="sc-dlfnbm eWIIxE"
        data-bg="#1d4289"
      >
        2299.1
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`Table simulate asset class column sort click 1`] = `
Array [
  <table
    aria-label="Financial Details"
    class="sc-bdfBwQ keewPp"
  >
    <thead>
      <tr
        class="sc-hKgILt eBPHrI"
      >
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="assertClass"
            id="assertClass"
          >
            Asset Class
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="ticker"
            id="ticker"
          >
            Ticker
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="price"
            id="price"
          >
            Price
          </button>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          BETA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          3791.37
        </td>
      </tr>
      <tr
        class="sc-gsTCUz ibRWgQ"
        data-theme="[object Object]"
      >
        <td>
          Credit
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          1168.46
        </td>
      </tr>
      <tr
        class="sc-gsTCUz jmeBcN"
        data-theme="[object Object]"
      >
        <td>
          Macro
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm kxGDBW"
          data-bg="#d3273e"
        >
          -1096.64
        </td>
      </tr>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          GAMMA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          2299.1
        </td>
      </tr>
    </tbody>
  </table>,
]
`;

exports[`Table simulate asset class column sort click 2`] = `
Array [
  <table
    aria-label="Financial Details"
    class="sc-bdfBwQ keewPp"
  >
    <thead>
      <tr
        class="sc-hKgILt eBPHrI"
      >
        <th>
          <button
            class="sc-eCssSg gzmARy"
            data-active="true"
            data-testid="assertClass"
            id="assertClass"
          >
            Asset Class
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="ticker"
            id="ticker"
          >
            Ticker
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="price"
            id="price"
          >
            Price
          </button>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          BETA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          3791.37
        </td>
      </tr>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          GAMMA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          2299.1
        </td>
      </tr>
      <tr
        class="sc-gsTCUz jmeBcN"
        data-theme="[object Object]"
      >
        <td>
          Macro
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm kxGDBW"
          data-bg="#d3273e"
        >
          -1096.64
        </td>
      </tr>
      <tr
        class="sc-gsTCUz ibRWgQ"
        data-theme="[object Object]"
      >
        <td>
          Credit
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          1168.46
        </td>
      </tr>
    </tbody>
  </table>,
]
`;

exports[`Table simulate price column sort click 1`] = `
Array [
  <table
    aria-label="Financial Details"
    class="sc-bdfBwQ keewPp"
  >
    <thead>
      <tr
        class="sc-hKgILt eBPHrI"
      >
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="assertClass"
            id="assertClass"
          >
            Asset Class
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="ticker"
            id="ticker"
          >
            Ticker
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="price"
            id="price"
          >
            Price
          </button>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          BETA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          3791.37
        </td>
      </tr>
      <tr
        class="sc-gsTCUz jmeBcN"
        data-theme="[object Object]"
      >
        <td>
          Macro
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm kxGDBW"
          data-bg="#d3273e"
        >
          -1096.64
        </td>
      </tr>
      <tr
        class="sc-gsTCUz ibRWgQ"
        data-theme="[object Object]"
      >
        <td>
          Credit
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          1168.46
        </td>
      </tr>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          GAMMA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          2299.1
        </td>
      </tr>
    </tbody>
  </table>,
]
`;

exports[`Table simulate price column sort click 2`] = `
Array [
  <table
    aria-label="Financial Details"
    class="sc-bdfBwQ keewPp"
  >
    <thead>
      <tr
        class="sc-hKgILt eBPHrI"
      >
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="assertClass"
            id="assertClass"
          >
            Asset Class
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="ticker"
            id="ticker"
          >
            Ticker
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gzmARy"
            data-active="true"
            data-testid="price"
            id="price"
          >
            Price
          </button>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          BETA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          3791.37
        </td>
      </tr>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          GAMMA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          2299.1
        </td>
      </tr>
      <tr
        class="sc-gsTCUz ibRWgQ"
        data-theme="[object Object]"
      >
        <td>
          Credit
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          1168.46
        </td>
      </tr>
      <tr
        class="sc-gsTCUz jmeBcN"
        data-theme="[object Object]"
      >
        <td>
          Macro
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm kxGDBW"
          data-bg="#d3273e"
        >
          -1096.64
        </td>
      </tr>
    </tbody>
  </table>,
]
`;

exports[`Table simulate ticker column sort click 1`] = `
Array [
  <table
    aria-label="Financial Details"
    class="sc-bdfBwQ keewPp"
  >
    <thead>
      <tr
        class="sc-hKgILt eBPHrI"
      >
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="assertClass"
            id="assertClass"
          >
            Asset Class
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="ticker"
            id="ticker"
          >
            Ticker
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="price"
            id="price"
          >
            Price
          </button>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          BETA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          3791.37
        </td>
      </tr>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          GAMMA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          2299.1
        </td>
      </tr>
      <tr
        class="sc-gsTCUz jmeBcN"
        data-theme="[object Object]"
      >
        <td>
          Macro
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm kxGDBW"
          data-bg="#d3273e"
        >
          -1096.64
        </td>
      </tr>
      <tr
        class="sc-gsTCUz ibRWgQ"
        data-theme="[object Object]"
      >
        <td>
          Credit
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          1168.46
        </td>
      </tr>
    </tbody>
  </table>,
]
`;

exports[`Table simulate ticker column sort click 2`] = `
Array [
  <table
    aria-label="Financial Details"
    class="sc-bdfBwQ keewPp"
  >
    <thead>
      <tr
        class="sc-hKgILt eBPHrI"
      >
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="assertClass"
            id="assertClass"
          >
            Asset Class
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gzmARy"
            data-active="true"
            data-testid="ticker"
            id="ticker"
          >
            Ticker
          </button>
        </th>
        <th>
          <button
            class="sc-eCssSg gYaMmc"
            data-active="false"
            data-testid="price"
            id="price"
          >
            Price
          </button>
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          BETA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          3791.37
        </td>
      </tr>
      <tr
        class="sc-gsTCUz jmeBcN"
        data-theme="[object Object]"
      >
        <td>
          Macro
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm kxGDBW"
          data-bg="#d3273e"
        >
          -1096.64
        </td>
      </tr>
      <tr
        class="sc-gsTCUz ibRWgQ"
        data-theme="[object Object]"
      >
        <td>
          Credit
        </td>
        <td>
          EPSILON
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          1168.46
        </td>
      </tr>
      <tr
        class="sc-gsTCUz kHDUxD"
        data-theme="[object Object]"
      >
        <td>
          Equities
        </td>
        <td>
          GAMMA
        </td>
        <td
          class="sc-dlfnbm eWIIxE"
          data-bg="#1d4289"
        >
          2299.1
        </td>
      </tr>
    </tbody>
  </table>,
]
`;
